kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/ShouldNotBeNullTest.kt

```diff
@@ -50,6 +50,16 @@ class ShouldNotBeNullTest : WordSpec() {
                .shouldStartWith("fo")
                .shouldEndWith("oo")
          }
+
+         "accept a block" {
+            val a: String? = "foo"
+
+            a shouldNotBeNull {
+               shouldHaveLength(3)
+               shouldStartWith("fo")
+               shouldEndWith("oo")
+            }
+         }
```

Comment:
It's currently possible to do either:

```kotlin
// Typical method chaining
            a.shouldNotBeNull()
              .shouldHaveLength(3)
              .shouldStartWith("fo")
              .shouldEndWith("oo")

// or, using Kotlin features
            a.run {
               shoudNotBeNull()
               shouldHaveLength(3)
               shouldStartWith("fo")
               shouldEndWith("oo")
            }
```

Are there other use-cases where this syntax provides a bigger difference?

Reply:
> Are there other use-cases where this syntax provides a bigger difference?

I don't thinks so. And you're right in saying that this is just syntactic sugar over using `.shouldNotBeNull().run {`. Mine is slightly shorter an (arguably) a tad more readable due to the infix notation.

Reply:
You can do `run` before even? `shouldNotBeNull` has a compiler contract which makes the compiler narrow the type to non-null if `shouldNotBeNull` passes

Reply:
> You can do `run` before even?

Not sure what you mean, can you give an example?

> `shouldNotBeNull` has a compiler contract

Yes, I'm aware of that. However, that contract does not work if `.shouldNotBeNull()` is used on a property that "is a public API property declared in a different module".

In my current case,

    receiveResponse.messages.shouldNotBeNull()
    receiveResponse.messages.forEach {  } // Smart cast impossible here.

does not work for that reason. Thus I prefer a syntax with a block in whose scope the property becomes non-nullable. So, 

    receiveResponse.messages.shouldNotBeNull().run { forEach {} }

works, but still

    receiveResponse.messages shouldNotBeNull { forEach {} }

reads a bit nicer, IMO.

Reply:
Thanks, that's a good point 🙂